### PR TITLE
ISPN-12193 Data distribution for cluster

### PR DESCRIPTION
**Pull request by jabolina**
https://issues.redhat.com/browse/ISPN-12193

* Added a new endpoint to the cluster endpoints that returns information
about each member in the cluster. We return the node's name, addresses,
memory available, and memory used.

* We extended the cluster stats to have an option for only the local
node and not the cluster-wide value. So we are exposing these metrics in
JMX too.

The endpoint is `GET /rest/v2/cluster?action=distribution` and the return is:

```json
[
  {
    "node_name": "NodeA",
    "node_addresses": [
      "127.0.0.1:39313"
    ],
    "memory_available": 466180016,
    "memory_used": 56010832
  },
  {
    "node_name": "NodeB",
    "node_addresses": [
      "127.0.0.1:47477"
    ],
    "memory_available": 467548568,
    "memory_used": 54642280
  }
]
```

The memory information is in bytes.

### CODE REVIEW

**Review by karesti — APPROVED** [review]

**Comment by tristantarrant** [comment]
What about per-cache distribution ? @wburns ?

File: core/src/main/java/org/infinispan/stats/impl/ClusterContainerStatsImpl.java
Comment (wburns) [inline core/src/main/java/org/infinispan/stats/impl/ClusterContainerStatsImpl.java:51]: This will not work in different JVM as the ArrayList is serialized to the remote node but no way to get the result back.
You can use the `submitConsumer` method which allows you to return a Map from the function which is serialized and returned back and passed to the TriConsumer which can collect the results.

File: server/rest/src/main/java/org/infinispan/rest/RestServer.java
Comment (wburns) [inline server/rest/src/main/java/org/infinispan/rest/RestServer.java:200]: We can't use members like this. It seems you are doing that to calculate when the consumers are all completed. However, this has a small window when a new member could join or leave that could cause it to miss an update or even never complete. 

Instead you should chain off the CompletionStage returned from the `submitConsumer` method to handle this.

Comment (jabolina) [reply server/rest/src/main/java/org/infinispan/rest/RestServer.java:200]: I've updated another place where I was doing the same thing. Thanks!!

File: server/rest/src/main/java/org/infinispan/rest/cachemanager/RestCacheManager.java
Comment (wburns) [inline server/rest/src/main/java/org/infinispan/rest/cachemanager/RestCacheManager.java:231]: Same thing, this should rethrow the exception instead of setting a future and return the stage returned from `thenApply`

File: server/rest/src/main/java/org/infinispan/rest/RestServer.java
Comment (jabolina) [reply server/rest/src/main/java/org/infinispan/rest/RestServer.java]: could not just `throw t` so we are throwing `CacheException` instead. Other places use an atomic reference to get the exception, we could that too here.

Comment (wburns) [reply server/rest/src/main/java/org/infinispan/rest/RestServer.java]: Sorry, just use `throw CompletableFutures.asCompletionException(t);`

**Comment by karesti** [comment]
@wburns @jabolina LGTM, I leave Will handle the merge of this PR

**Review by karesti — APPROVED** [review]

**Review by wburns — COMMENTED** [review]
LGTM, however it appears there is a related failure https://ci.infinispan.org/job/Infinispan/job/PR-10171/5/testReport/org.infinispan.rest.resources/CacheResourceV2Test/testCacheAvailability_security_false__protocol_HTTP_11__ssl_false_/ as it has failed 4 times in a row now.

File: server/rest/src/main/java/org/infinispan/rest/RestServer.java
Comment (pruivo) [inline server/rest/src/main/java/org/infinispan/rest/RestServer.java:208]: nitpick. this should be in `ClusterResource` class.

File: server/rest/src/main/java/org/infinispan/rest/RestServer.java
Comment (pruivo) [inline server/rest/src/main/java/org/infinispan/rest/RestServer.java:210]: nitpick. if members are null, return the local memory.

File: server/rest/src/main/java/org/infinispan/rest/resources/ClusterResource.java
Comment (pruivo) [inline server/rest/src/main/java/org/infinispan/rest/resources/ClusterResource.java:105]: nitpick: `CompletableFuture.supplyAsync` not required.

File: commons/all/src/main/java/org/infinispan/commons/marshall/ProtoStreamTypeIds.java
Comment (pruivo) [inline commons/all/src/main/java/org/infinispan/commons/marshall/ProtoStreamTypeIds.java:147]: unused

**Review by pruivo — COMMENTED** [review]
LGTM, just some minor comments.

I'm wondering why we are naming it "distribution" if only reports the memory...

File: server/rest/src/main/java/org/infinispan/rest/resources/ClusterResource.java
Comment (jabolina) [reply server/rest/src/main/java/org/infinispan/rest/resources/ClusterResource.java]: Was unsure how to act, should just call `this.clusterDistribution()`?

Comment (pruivo) [reply server/rest/src/main/java/org/infinispan/rest/resources/ClusterResource.java]: yes. the `clusterDistribution()` method does not block since only submits works to the cluster executor.

**Comment by karesti** [comment]
@jabolina is this ready? @pruivo @wburns

**Review by pruivo — APPROVED** [review]

**Comment by jabolina** [comment]
I've put it on hold until I check the use of the security actions, as per #10234 #10232

**Comment by jabolina** [comment]
I'll keep the label until we merge #10234 as that will cause conflict with the changes here.

**Comment by jabolina** [comment]
Guess we're good. Waiting CI.

**Comment by pruivo** [comment]
merged! thanks @jabolina !

**Pull request merged**
